### PR TITLE
Deleting executions

### LINKED ISSUE #383: Deleting executions
We must have functionality to remove old executions

### PR DESCRIPTION
**Pull request by Cheshiriks**
What's done:
* added functionality to delete executions
Closes #383

### CODE REVIEW

File: save-backend/src/main/kotlin/org/cqfn/save/backend/controllers/ExecutionController.kt [hunk 1]
```diff
@@ -124,6 +130,28 @@ class ExecutionController(private val executionService: ExecutionService,
...
+    /**
+     * Delete all executions by project name and project owner
+     *
+     * @param name name of project
+     * @param owner owner of project
+     * @return ResponseEntity
+     * @throws ResponseStatusException
+     */
+    @PostMapping("/deleteExecution")
+    fun deleteExecutionForProject(@RequestParam name: String, @RequestParam owner: String): ResponseEntity<String>? {
```
Comment (petertrr) [inline save-backend/src/main/kotlin/org/cqfn/save/backend/controllers/ExecutionController.kt:142]: I think we should be able to delete executions selectively, by ID or a list of IDs. Clearing the history completely from the frontend is needed not as often.

File: save-backend/src/main/kotlin/org/cqfn/save/backend/controllers/ExecutionController.kt [hunk 2]
```diff
@@ -124,6 +131,45 @@ class ExecutionController(private val executionService: ExecutionService,
...
+    /**
+     * Delete all executions by project name and project owner
+     *
+     * @param name name of project
+     * @param owner owner of project
+     * @return ResponseEntity
+     * @throws ResponseStatusException
+     */
+    @PostMapping("/deleteAllExecution")
+    fun deleteExecutionForProject(@RequestParam name: String, @RequestParam owner: String): ResponseEntity<String>? {
```
Comment (petertrr) [inline save-backend/src/main/kotlin/org/cqfn/save/backend/controllers/ExecutionController.kt:143]:
```suggestion
    fun deleteExecutionForProject(@RequestParam name: String, @RequestParam owner: String): ResponseEntity<String> {
```
Why is it nullable?

File: save-backend/src/main/kotlin/org/cqfn/save/backend/controllers/ExecutionController.kt [hunk 3]
```diff
@@ -124,6 +131,45 @@ class ExecutionController(private val executionService: ExecutionService,
...
+        val projectId = projectService.findByNameAndOwner(name, owner)?.id
+
+        try {
+            testExecutionService.deleteTestExecutionWithProjectId(projectId)
+            agentService.deleteAgentWithProjectId(projectId)
```
Comment (petertrr) [inline save-backend/src/main/kotlin/org/cqfn/save/backend/controllers/ExecutionController.kt:148]: Also need to delete agent statuses?

File: save-backend/src/main/kotlin/org/cqfn/save/backend/service/AgentService.kt [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+     */
+    internal fun deleteAgentWithProjectId(projectId: Long?) {
+        val agents = agentRepository.findAll { root, _, cb ->
+            cb.equal(root.get<Execution>("execution").get<Project>("project").get<Long>("id"), projectId)
+        }
+
+        agents.forEach {
+            agentRepository.delete(it)
+        }
+        agentRepository.flush()
```
Comment (petertrr) [inline save-backend/src/main/kotlin/org/cqfn/save/backend/service/AgentService.kt:24]: Why do you need `flush`? All deletions will happen in a single transaction

File: save-backend/src/main/kotlin/org/cqfn/save/backend/service/TestExecutionService.kt [hunk 5]
```diff
@@ -82,6 +87,36 @@ class TestExecutionService(private val testExecutionRepository: TestExecutionRep
     internal fun getTestExecutionsCount(executionId: Long) = testExecutionRepository
         .countByExecutionId(executionId)
 
+    /**
+     * @param projectId
+     */
+    internal fun deleteTestExecutionWithProjectId(projectId: Long?) {
+        val testExecutions = testExecutionRepository.findAll { root, cq, cb ->
```
Comment (petertrr) [inline save-backend/src/main/kotlin/org/cqfn/save/backend/service/TestExecutionService.kt:94]: Same, may be easier and more readable with spring-data method naming and runtime proxies

File: save-backend/src/test/kotlin/org/cqfn/save/backend/controller/ExecutionControllerTest.kt [hunk 6]
```diff
@@ -118,6 +119,29 @@ class ExecutionControllerTest {
         assertTrue(databaseData.any { it.status == execution.status && it.startTime == testLocalDateTime })
     }
 
+    @Test
+    @Disabled  // fixme: should rollback after committing in the db
```
Comment (petertrr) [inline save-backend/src/test/kotlin/org/cqfn/save/backend/controller/ExecutionControllerTest.kt:123]: Okay, are you going to fix this fixme?

File: save-frontend/src/main/kotlin/org/cqfn/save/frontend/components/views/HistoryView.kt [hunk 7]
```diff
@@ -166,6 +263,71 @@ class HistoryView : RComponent<HistoryProps, State>() {
...
+
+    private fun deleteExecutionBuilder(executionIds: List<Long>) {
+        val headers = Headers().also {
+            it.set("Accept", "application/json")
+            it.set("Content-Type", "application/json")
+        }
+        GlobalScope.launch {
+            responseFromDeleteExecutions =
+                    post("${window.location.origin}/deleteExecution?executionIds=${executionIds.joinToString(",")}", headers, undefined)
+        }.invokeOnCompletion {
```
Comment (petertrr) [inline save-frontend/src/main/kotlin/org/cqfn/save/frontend/components/views/HistoryView.kt:316]: No need to use `inokeOnCompletion` here, everything can be done inside `launch`

**Review by petertrr — CHANGES_REQUESTED** [review]

File: save-backend/src/main/kotlin/org/cqfn/save/backend/controllers/ExecutionController.kt [hunk 8]
```diff
@@ -124,6 +131,45 @@ class ExecutionController(private val executionService: ExecutionService,
...
+     * @throws ResponseStatusException
+     */
+    @PostMapping("/deleteExecution")
+    fun deleteExecutionsByExecutionIds(@RequestParam executionIds: List<Long>): ResponseEntity<String>? {
+        try {
+            testExecutionService.deleteTestExecutionByExecutionIds(executionIds)
+            agentService.deleteAgentByExecutionIds(executionIds)
+            executionService.deleteExecutionByIds(executionIds)
+        } catch (e: IllegalArgumentException) {
+            throw ResponseStatusException(HttpStatus.NOT_FOUND, "Failed to delete executions for the following reason: ${e.message}")
```
[the remark this answers is not in the document]
Comment (Cheshiriks) [reply save-backend/src/main/kotlin/org/cqfn/save/backend/controllers/ExecutionController.kt:168]: IllegalArgumentException throw of `delete` method if entity or id not found

File: save-backend/src/main/kotlin/org/cqfn/save/backend/controllers/ExecutionController.kt (hunk 3 above)
Comment (Cheshiriks) [reply save-backend/src/main/kotlin/org/cqfn/save/backend/controllers/ExecutionController.kt:148]: Yes. And need add foreign key on agent_id in agent_status

File: save-backend/src/main/kotlin/org/cqfn/save/backend/controllers/ExecutionController.kt [hunk 9]
```diff
@@ -124,6 +131,45 @@ class ExecutionController(private val executionService: ExecutionService,
...
+        }
+        return ResponseEntity.status(HttpStatus.OK).build()
+    }
+
+    /**
+     * @param executionIds list of ids
+     * @return ResponseEntity
+     * @throws ResponseStatusException
+     */
+    @PostMapping("/deleteExecution")
```
[the remark this answers is not in the document]
Comment (Cheshiriks) [reply save-backend/src/main/kotlin/org/cqfn/save/backend/controllers/ExecutionController.kt:161]: Fixed

File: save-backend/src/main/kotlin/org/cqfn/save/backend/service/AgentService.kt [hunk 10]
```diff
@@ -0,0 +1,36 @@
...
+/**
+ * Service for agent
+ */
+@Service
+class AgentService(private val agentRepository: AgentRepository) {
+    /**
+     * @param projectId
+     */
+    internal fun deleteAgentWithProjectId(projectId: Long?) {
+        val agents = agentRepository.findAll { root, _, cb ->
```
[the remark this answers is not in the document]
Comment (Cheshiriks) [reply save-backend/src/main/kotlin/org/cqfn/save/backend/service/AgentService.kt:17]: Fixed

File: save-backend/src/main/kotlin/org/cqfn/save/backend/controllers/ExecutionController.kt (hunk 2 above)
Comment (Cheshiriks) [reply save-backend/src/main/kotlin/org/cqfn/save/backend/controllers/ExecutionController.kt:143]: Fixed

File: save-backend/src/main/kotlin/org/cqfn/save/backend/controllers/ExecutionController.kt (hunk 8 above)
[the remark this answers is not in the document]
Comment (petertrr) [reply save-backend/src/main/kotlin/org/cqfn/save/backend/controllers/ExecutionController.kt:168]: Okay, pls also add `log.warn` here with `executionIds` and any other useful info

File: save-frontend/src/main/kotlin/org/cqfn/save/frontend/components/views/HistoryView.kt (hunk 7 above)
Comment (Cheshiriks) [reply save-frontend/src/main/kotlin/org/cqfn/save/frontend/components/views/HistoryView.kt]: Fixed

File: save-backend/src/main/kotlin/org/cqfn/save/backend/service/TestExecutionService.kt (hunk 5 above)
Comment (Cheshiriks) [reply save-backend/src/main/kotlin/org/cqfn/save/backend/service/TestExecutionService.kt:94]: Yes. Fixed

File: save-backend/src/main/kotlin/org/cqfn/save/backend/service/AgentService.kt (hunk 4 above)
Comment (Cheshiriks) [reply save-backend/src/main/kotlin/org/cqfn/save/backend/service/AgentService.kt]: Ok. Deleted

File: save-backend/src/main/kotlin/org/cqfn/save/backend/controllers/ExecutionController.kt (hunk 3 above)
Comment (Cheshiriks) [reply save-backend/src/main/kotlin/org/cqfn/save/backend/controllers/ExecutionController.kt]: Added

File: save-backend/src/main/kotlin/org/cqfn/save/backend/controllers/ExecutionController.kt [hunk 11]
```diff
@@ -139,15 +141,17 @@ class ExecutionController(private val executionService: ExecutionService,
...
-            agentService.deleteAgentWithProjectId(projectId)
-            executionService.deleteExecutionByProjectNameAndProjectOwner(name, owner)
+            projectService.findByNameAndOwner(name, owner)?.id?.let {
+                testExecutionService.deleteTestExecutionWithProjectId(it)
+                agentStatusService.deleteAgentStatusWithProjectId(it)
+                agentService.deleteAgentWithProjectId(it)
+                executionService.deleteExecutionByProjectNameAndProjectOwner(name, owner)
+            }
         } catch (e: IllegalArgumentException) {
+            log.warn("Could not find the project with name: $name and owner: $owner or related objects")
```
Comment (petertrr) [inline save-backend/src/main/kotlin/org/cqfn/save/backend/controllers/ExecutionController.kt:154]:
```suggestion
            log.warn("Could not find the project with name: $name and owner: $owner or related objects", e)
```

File: save-backend/src/main/kotlin/org/cqfn/save/backend/controllers/ExecutionController.kt [hunk 12]
```diff
@@ -139,15 +141,17 @@ class ExecutionController(private val executionService: ExecutionService,
...
-    fun deleteExecutionForProject(@RequestParam name: String, @RequestParam owner: String): ResponseEntity<String>? {
-        val projectId = projectService.findByNameAndOwner(name, owner)?.id
-
+    @PostMapping("/execution/deleteAll")
+    fun deleteExecutionForProject(@RequestParam name: String, @RequestParam owner: String): ResponseEntity<String> {
         try {
-            testExecutionService.deleteTestExecutionWithProjectId(projectId)
-            agentService.deleteAgentWithProjectId(projectId)
-            executionService.deleteExecutionByProjectNameAndProjectOwner(name, owner)
+            projectService.findByNameAndOwner(name, owner)?.id?.let {
```
Comment (petertrr) [inline save-backend/src/main/kotlin/org/cqfn/save/backend/controllers/ExecutionController.kt:147]: Currently you are silently ignoring null values here. If project is not found (i.e. `find...` returns null), you should also return 404. You can achieve this by, for example, wrapping `find` in `requireNotNull`, which too throws `IllegalArgumentException`. And if project is found, then its `id` is guaranteed to be not null, so you can `!!` it.

File: save-backend/src/main/kotlin/org/cqfn/save/backend/controllers/ExecutionController.kt [hunk 13]
```diff
@@ -158,13 +162,15 @@ class ExecutionController(private val executionService: ExecutionService,
...
-    @PostMapping("/deleteExecution")
+    @PostMapping("/execution/delete")
     fun deleteExecutionsByExecutionIds(@RequestParam executionIds: List<Long>): ResponseEntity<String>? {
         try {
             testExecutionService.deleteTestExecutionByExecutionIds(executionIds)
+            agentStatusService.deleteAgentStatusWithExecutionIds(executionIds)
             agentService.deleteAgentByExecutionIds(executionIds)
             executionService.deleteExecutionByIds(executionIds)
         } catch (e: IllegalArgumentException) {
+            log.warn("Could not find the following executions: $executionIds or related objects")
```
Comment (petertrr) [inline save-backend/src/main/kotlin/org/cqfn/save/backend/controllers/ExecutionController.kt:173]:
```suggestion
            log.warn("Could not find the following executions: $executionIds or related objects", e)
```

File: save-backend/src/main/kotlin/org/cqfn/save/backend/service/AgentService.kt [hunk 14]
```diff
@@ -12,26 +10,17 @@ import org.springframework.stereotype.Service
 class AgentService(private val agentRepository: AgentRepository) {
     /**
      * @param projectId
+     * @return Unite
```
Comment (petertrr) [inline save-backend/src/main/kotlin/org/cqfn/save/backend/service/AgentService.kt:13]:
```suggestion
     * @return Unit
```

File: save-backend/src/test/kotlin/org/cqfn/save/backend/controller/DeleteEntitiesTest.kt [hunk 15]
```diff
@@ -0,0 +1,49 @@
...
+        webClient.post()
+            .uri("/execution/delete?executionIds=$ids")
+            .contentType(MediaType.APPLICATION_JSON)
+            .exchange()
+            .expectStatus()
+            .isOk
+    }
+
+    @Test
+    @Disabled  // fixme: should rollback after committing in the db
```
Comment (petertrr) [inline save-backend/src/test/kotlin/org/cqfn/save/backend/controller/DeleteEntitiesTest.kt:40]: `@Disabled` accepts reason as an argument

File: save-backend/src/main/kotlin/org/cqfn/save/backend/controllers/ExecutionController.kt (hunk 11 above)
Comment (Cheshiriks) [reply save-backend/src/main/kotlin/org/cqfn/save/backend/controllers/ExecutionController.kt]: Fixed

File: save-backend/src/main/kotlin/org/cqfn/save/backend/controllers/ExecutionController.kt (hunk 12 above)
Comment (Cheshiriks) [reply save-backend/src/main/kotlin/org/cqfn/save/backend/controllers/ExecutionController.kt]: Fixed

File: save-backend/src/main/kotlin/org/cqfn/save/backend/controllers/ExecutionController.kt (hunk 13 above)
Comment (Cheshiriks) [reply save-backend/src/main/kotlin/org/cqfn/save/backend/controllers/ExecutionController.kt]: Fixed

File: save-backend/src/main/kotlin/org/cqfn/save/backend/service/AgentService.kt (hunk 14 above)
Comment (Cheshiriks) [reply save-backend/src/main/kotlin/org/cqfn/save/backend/service/AgentService.kt]: Fixed all

File: save-backend/src/test/kotlin/org/cqfn/save/backend/controller/DeleteEntitiesTest.kt (hunk 15 above)
Comment (Cheshiriks) [reply save-backend/src/test/kotlin/org/cqfn/save/backend/controller/DeleteEntitiesTest.kt]: Fixed

File: db/v-2/tables/agent-status.xml [hunk 16]
```diff
@@ -0,0 +1,16 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<databaseChangeLog
+        xmlns="http://www.liquibase.org/xml/ns/dbchangelog"
+        xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+        xsi:schemaLocation="http://www.liquibase.org/xml/ns/dbchangelog
+                      http://www.liquibase.org/xml/ns/dbchangelog/dbchangelog-3.8.xsd">
+
```
Comment (akuleshov7) [inline db/v-2/tables/agent-status.xml:7]: Why you are adding it. We did not have a foreign key?

File: db/v-2/tables/agent-status.xml [hunk 17]
```diff
@@ -0,0 +1,16 @@
...
+    <changeSet id="agent-status-addFK" author="frolov" context="dev or prod">
```
Comment (akuleshov7) [inline db/v-2/tables/agent-status.xml:8]: Why not add-fk

File: save-backend/src/main/kotlin/org/cqfn/save/backend/controllers/ExecutionController.kt (hunk 8 above)
[the remark this answers is not in the document]
Comment (akuleshov7) [reply save-backend/src/main/kotlin/org/cqfn/save/backend/controllers/ExecutionController.kt:175]: I actually like Petr’s comment. Not found is usually related to a page URL.
Have you also checked the exception. Is there any information about the TABLE where we did not find the executionId

File: save-backend/src/main/kotlin/org/cqfn/save/backend/service/TestExecutionService.kt [hunk 18]
```diff
@@ -82,6 +87,31 @@ class TestExecutionService(private val testExecutionRepository: TestExecutionRep
...
+            val executionSq: Subquery<Long> = cq.subquery(Long::class.java)
+            val executionRoot: Root<Execution> = executionSq.from(Execution::class.java)
```
Comment (akuleshov7) [inline save-backend/src/main/kotlin/org/cqfn/save/backend/service/TestExecutionService.kt:96]: Vlad, this is not readable, better to use a native query language

**Review by akuleshov7 — CHANGES_REQUESTED** [review]
Need to think about natural query

File: db/v-2/tables/agent-status.xml (hunk 16 above)
Comment (Cheshiriks) [reply db/v-2/tables/agent-status.xml:7]: No. We didn't have a foreign key and we could delete agent and didn't delete agent status

File: db/v-2/tables/agent-status.xml (hunk 17 above)
Comment (Cheshiriks) [reply db/v-2/tables/agent-status.xml:8]: Do you mean name id? agent-status-add-FK ?

File: save-backend/src/main/kotlin/org/cqfn/save/backend/service/TestExecutionService.kt (hunk 18 above)
Comment (Cheshiriks) [reply save-backend/src/main/kotlin/org/cqfn/save/backend/service/TestExecutionService.kt:96]: It won't work. Entity TestExecution don't have link on entity Execution. It have executionId Long

Comment (petertrr) [reply save-backend/src/main/kotlin/org/cqfn/save/backend/service/TestExecutionService.kt:96]: Andrey means native SQL query, see [here](https://www.baeldung.com/spring-data-jpa-query).

Comment (Cheshiriks) [reply save-backend/src/main/kotlin/org/cqfn/save/backend/service/TestExecutionService.kt:96]: > Andrey means native SQL query, see [here](https://www.baeldung.com/spring-data-jpa-query).

No. It won't work for the same reason.

**Review by petertrr — APPROVED** [review]

**Review by akuleshov7 — APPROVED** [review]

**Pull request merged**